R/spatial_nndm_cv.R

```diff
@@ -0,0 +1,207 @@
+        i = "Reproject `prediction_sites` and `data` to share a CRS to avoid this warning."
+      )
+    )
+    prediction_sites <- sf::st_transform(prediction_sites, sf::st_crs(data))
+  }
+
+  cv_att <- list(
+    prediction_sites = prediction_sites,
+    prediction_sample_size = prediction_sample_size,
+    autocorrelation_range = autocorrelation_range,
+    min_analysis_proportion = min_analysis_proportion,
+    ...
+  )
+
+  pred_geometry <- unique(sf::st_geometry_type(prediction_sites))
+  if (length(pred_geometry) > 1 || pred_geometry != "POINT") {
+    prediction_sites <- sf::st_sample(
+      x = sf::st_as_sfc(sf::st_bbox(prediction_sites)),
+      size = prediction_sample_size,
+      ...
+    )
+  }
+
+  if (is.null(autocorrelation_range)) {
+    bbox <- sf::st_bbox(prediction_sites)
+
+    autocorrelation_range <- sf::st_distance(
+      sf::st_as_sf(
+        data.frame(
+          lon = bbox[c("xmin", "xmax")],
+          lat = bbox[c("ymin", "ymax")]
+        ),
+        coords = c("lon", "lat"),
+        crs = sf::st_crs(prediction_sites)
+      )
+    )[2]
+  }
+
+  nn_prediction <- apply(sf::st_distance(prediction_sites, data), 1, min)
+
+  distance_matrix <- sf::st_distance(data)
+
+  # We've enforced that prediction_sites and data are in the same CRS;
+  # therefore nearest_neighbors and distance_matrix are in the same units
+  # Force autocorrelation_range into the same units:
+  units(autocorrelation_range) <- units(distance_matrix)
+
+  # We're guaranteed to be working in one set of units now,
+  # which means we should be able to drop units entirely at this point
+  # (which should make some of the logic here easier)
+  units(autocorrelation_range) <- NULL
+  units(distance_matrix) <- NULL
+
+  diag(distance_matrix) <- NA
+  nn_training <- apply(distance_matrix, 1, min, na.rm = TRUE)
+
+  indices <- list(
+    distance = min(nn_training),
+    row = which.min(nn_training)[1]
+  )
+  indices$col <- which.min(distance_matrix[indices$row, ])
+
+  n_training <- nrow(data)
+
+  while (indices$distance <= autocorrelation_range) {
+    # Proportion of training data with a neighbor in training
+    # closer than indices$distance if we removed one additional point
+    # (hence 1 / n_training)
+    prop_close_training <-
+      mean(nn_training <= indices$distance) - (1 / n_training)
+    # Proportion of training data with a neighbor in prediction data
+    # closer than indices$distance
+    prop_close_prediction <- mean(nn_prediction <= indices$distance)
+
+    # How much data remains in analysis sets?
+    prop_remaining <- sum(!is.na(distance_matrix[indices$row, ])) / n_training
+
+    if ((prop_close_training >= prop_close_prediction) &
+      (prop_remaining > min_analysis_proportion)) {
+      distance_matrix[indices$row, indices$col] <- NA
+
+      nn_training <- apply(distance_matrix, 1, min, na.rm = TRUE)
+
+      indices <- update_indices(indices, nn_training, distance_matrix, `>=`)
```

Comment:
this corresponds to finding the next-nearest pair, right? That function name is also rather generic and I think it would be easier to read if it was linked more with the domain (= the cross-validation method of nearest neighbor).